org.osgi.test.junit5/src/test/java/org/osgi/test/junit5/config/ConfigAnnotationTest.java

```diff
@@ -0,0 +1,171 @@
+	private static BundleContext						bundleContext				= null;
+	private static ConfigurationAdmin					ca							= null;
+
+	@BeforeAll
+	public static void beforeEach() {
+
+		bundleContext = FrameworkUtil.getBundle(ConfigAnnotationTest.class)
+			.getBundleContext();
+
+		sref = bundleContext.getServiceReference(ConfigurationAdmin.class);
+		if (sref == null) {
+			throw new IllegalStateException("The configuration admin service cannot be found.");
+		}
+
+		ca = bundleContext.getService(sref);
+		if (ca == null) {
+			throw new IllegalStateException("The configuration admin service cannot be found.");
+		}
+
+	}
+
+	@AfterAll
+	public static void afterEach() {
+		BundleContext bundleContext = FrameworkUtil.getBundle(ConfigAnnotationTest.class)
+			.getBundleContext();
+		if (bundleContext != null && sref != null) {
+			bundleContext.ungetService(sref);
+		}
+
+	}
+
+	private List<org.osgi.service.cm.Configuration> getConfigsByServicePid(String pid) throws Exception {
+		return getConfigsBy(Constants.SERVICE_PID, pid);
+	}
+
+	private List<org.osgi.service.cm.Configuration> getConfigsByFactoryServicePid(String pid) throws Exception {
+		return getConfigsBy("service.factoryPid", pid);
+	}
+
+	private List<org.osgi.service.cm.Configuration> getConfigsBy(String key, String value) throws Exception {
+		String filter = String.format("(%s=%s)", key, value);
+		org.osgi.service.cm.Configuration[] configurations = ca.listConfigurations(filter);
+		if (configurations == null) {
+			return Collections.EMPTY_LIST;
+		}
+		return Arrays.asList(configurations);
+
+	}
+
+	// START TESTS
+	@Configuration(pid = NONSTATIC_CONFIGURATION_PID)
+	org.osgi.service.cm.Configuration nonStaticConfiguration;
+
+	@Test
+	public void testFieldConfiguration() throws Exception {
+
+		List<org.osgi.service.cm.Configuration> cs = getConfigsByServicePid(NONSTATIC_CONFIGURATION_PID);
+		assertEquals(1, cs.size());
+		assertEquals(cs.get(0), nonStaticConfiguration);
+
+	}
+
+	static final String PARAM_PID = "param.pid";
+
+	@Test
+	public void testParameterConfiguration(@Configuration(pid = PARAM_PID, properties = {
+		@ConfigEntry(key = "bar", value = "foo")
+	}) org.osgi.service.cm.Configuration configuration) throws Exception {
+
+		List<org.osgi.service.cm.Configuration> cs = getConfigsByServicePid(PARAM_PID);
+		assertEquals(1, cs.size());
+		assertEquals(cs.get(0), configuration);
+
+		assertNull(configuration.getProperties()
+			.get("foo"));
+		String foo = (String) configuration.getProperties()
+			.get("bar");
+		assertEquals("foo", foo);
+	}
+
+	@Test
+	public void testParameterConfiguration2(@Configuration(pid = PARAM_PID, properties = {
+		@ConfigEntry(key = "foo", value = "bar")
+	}) org.osgi.service.cm.Configuration configuration) throws Exception {
+
+		List<org.osgi.service.cm.Configuration> cs = getConfigsByServicePid(PARAM_PID);
+		assertEquals(1, cs.size());
+		assertEquals(cs.get(0), configuration);
```

Comment:
AssertJ is much better at this than standard JUnit 5 assertions: `assertThat(cs).containsExactly(configuration)` Not only is the code more compact and readable, the diagnostic message (if the assertion fails) is much better too.

I highly recommend using AssertJ for these tests - we've used it for most of the unit tests in this project. Additionally, if you're going to be doing multiple assertions per test, you should look at using soft assertions (perhaps with the `SoftAssertionsExtension` to make your life easier).